File: ios/Sources/UI/ReactViewFromJsonResolver.swift
/*
 * This file is part of the Scandit Data Capture SDK
 *
 * Copyright (C) 2025- Scandit AG. All rights reserved.
 */

import Foundation
import React
import ScanditDataCaptureCore
import ScanditFrameworksBarcode
import ScanditFrameworksCore
import UIKit

public class ReactViewFromJsonResolver: ViewFromJsonResolver {
    private weak var container: AdvancedOverlayContainer?

    public init(container: AdvancedOverlayContainer) {
        self.container = container
    }

    public func getView(viewJson: String?) -> UIView? {
        guard let viewJson = viewJson, let jsonData = viewJson.data(using: .utf8) else {
            return nil
        }

        guard let configuration = try? JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any]
        else {
            return nil
        }

        guard let jsView = try? JSView(with: configuration) else {
            return nil
        }

        return container?.rootViewWith(jsView: jsView)
    }

    public func getViewFromBytes(
        advancedOverlayViewPool: AdvancedOverlayViewCache,
        viewIdentifier: String,
        viewBytes: Data?
    ) -> UIView? {
        nil
    }
}

File: android/src/main/kotlin/com/scandit/datacapture/reactnative/barcode/ui/ReactViewFromJsonResolver.kt
/*
 * This file is part of the Scandit Data Capture SDK
 *
 * Copyright (C) 2025- Scandit AG. All rights reserved.
 */

import android.view.View
import com.facebook.react.bridge.ReactApplicationContext
import com.scandit.datacapture.frameworks.core.ui.ViewFromJsonResolver
import com.scandit.datacapture.frameworks.core.utils.AdvancedOverlayViewCache
import com.scandit.datacapture.reactnative.barcode.batch.nativeViewFromJson

class ReactViewFromJsonResolver(
    private val reactContext: ReactApplicationContext,
) : ViewFromJsonResolver {
    override fun getView(
        advancedOverlayViewPool: AdvancedOverlayViewCache,
        viewIdentifier: String,
        viewJson: String?
    ): View? {
        // Reuse the pooled view for this tracked barcode if one already exists.
        // The barcode SDK re-requests the view for a tracked barcode on every
        // session update; recreating a fresh ScanditReactRootView (a full
        // startReactApplication / ReactSurface mount) each time makes the bubble
        // flicker. Mirror DefaultViewFromJsonResolver, which reuses by identifier
        // via the pool. The caller (BarcodeBatchModule.setViewForTrackedBarcode)
        // populates the pool after creation. See SDC-32212.
        advancedOverlayViewPool.getView(viewIdentifier)?.let { return it }

        val activity = reactContext.currentActivity ?: return null
        return nativeViewFromJson(activity, viewJson)
    }
}
